hazelcast/include/hazelcast/client/proxy/SerializingProxy.h

```diff
@@ -92,41 +93,46 @@ class HAZELCAST_API SerializingProxy
         return to_data<T>(&object);
     }
 
-    template<typename T>
-    inline boost::optional<T> to_object(const serialization::pimpl::data& data)
-    {
-        return serialization_service_.template to_object<T>(data);
-    }
-
-    template<typename T>
-    inline boost::optional<T> to_object(const serialization::pimpl::data* data)
-    {
-        if (!data) {
-            return boost::none;
-        } else {
-            return to_object<T>(*data);
-        }
-    }
-
-    template<typename T>
-    inline boost::optional<T> to_object(
-      std::unique_ptr<serialization::pimpl::data>&& data)
-    {
-        return to_object<T>(data.get());
-    }
-
-    template<typename T>
-    inline boost::optional<T> to_object(
-      std::unique_ptr<serialization::pimpl::data>& data)
+    template<typename K, typename V>
+    inline boost::future<std::unordered_map<K, V>> to_object_map(
+      std::vector<boost::future<EntryVector>>& futures)
     {
-        return to_object<T>(data.get());
+        return boost::when_all(futures.begin(), futures.end())
+          .then(boost::launch::sync,
+                [this](
+                  boost::future<boost::csbl::vector<boost::future<EntryVector>>>
+                    results_data) {
+                    std::unordered_map<K, V> result;
+                    for (auto& entryVectorFuture : results_data.get()) {
+                        for (auto& entry : entryVectorFuture.get()) {
+                            auto val = to_object<V>(entry.second);
+                            // it is guaranteed that all values are non-null
+                            assert(val.has_value());
+                            result.emplace(to_object<K>(entry.first).value(),
+                                           std::move(val.value()));
+                        }
+                    }
+                    return result;
+                });
     }
 
     template<typename T>
-    inline boost::optional<T> to_object(
-      const boost::optional<serialization::pimpl::data>& data)
+    inline boost::future<size_t> to_object_drain(
+      boost::future<std::vector<serialization::pimpl::data>> f,
+      std::vector<T>& elements)
     {
-        return to_object<T>(data.get_ptr());
+        return f.then(
+          boost::launch::sync,
+          [&elements,
+           this](boost::future<std::vector<serialization::pimpl::data>> f) {
+              auto datas = f.get();
+              auto size = datas.size();
+              elements.reserve(elements.size() + size);
```

Comment:
There may be a bug here. According to javadoc, it says `Removes all available elements from this queue and adds them to the given collection`
Can you double check on this?

Reply:
I double checked, this is the correct behaviour.

Reply:
Note that we are not changing the behaviour. This is just more optimized version where we resize the elements vector given from user once. In existing impl, push_back my had to resize more than once.